### PR TITLE
Add the test_bench userbenchmark

### PR DESCRIPTION
**Pull request by xuzhao9**
The plan is to use the `test_bench` userbenchmark to deprecate `test_bench.py`.

It supports:
1. Running multiple models, each model in its subprocess.
2. Running with extra args such as `--torchdynamo inductor`. 
3. `--debug` option: save the output of subprocess in the `output-%Y%m%d%H%M%S` directory.

Test plan:

```
$ python run_benchmark.py test_bench llama_v2_7b_16h -d cuda -t eval --accuracy --debug
Running TorchBenchModelConfig(name='llama_v2_7b_16h', test='eval', device='cuda', batch_size=None, extra_args=['--accuracy'], extra_env=None, output_dir=PosixPath("/data/users/xzhao9/git/benchmark/.userbenchmark/test_bench/output-20231122193846/model=llama_v2_7b_16h, test=eval, device=cuda, bs=None, extra_args=['--accuracy']")) ...[done]
```

The error log is saved to the log file: `test_bench/output-20231122193846/model=llama_v2_7b_16h, test=eval, device=cuda, bs=None, extra_args=['--accuracy']/stderr.log`:

```
  File "/home/xzhao9/.conda/envs/py38/lib/python3.8/site-packages/torch/nn/modules/module.py", line 1510, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
  File "/home/xzhao9/.conda/envs/py38/lib/python3.8/site-packages/torch/nn/modules/module.py", line 1519, in _call_impl
    return forward_call(*args, **kwargs)
  File "/home/xzhao9/.conda/envs/py38/lib/python3.8/site-packages/transformers/models/llama/modeling_llama.py", line 424, in forward
    hidden_states, self_attn_weights, present_key_value = self.self_attn(
  File "/home/xzhao9/.conda/envs/py38/lib/python3.8/site-packages/torch/nn/modules/module.py", line 1510, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
  File "/home/xzhao9/.conda/envs/py38/lib/python3.8/site-packages/torch/nn/modules/module.py", line 1519, in _call_impl
    return forward_call(*args, **kwargs)
  File "/home/xzhao9/.conda/envs/py38/lib/python3.8/site-packages/transformers/models/llama/modeling_llama.py", line 321, in forward
    query_states = self.q_proj(hidden_states)
  File "/home/xzhao9/.conda/envs/py38/lib/python3.8/site-packages/torch/nn/modules/module.py", line 1510, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
  File "/home/xzhao9/.conda/envs/py38/lib/python3.8/site-packages/torch/nn/modules/module.py", line 1519, in _call_impl
    return forward_call(*args, **kwargs)
  File "/home/xzhao9/.conda/envs/py38/lib/python3.8/site-packages/torch/nn/modules/linear.py", line 116, in forward
    return F.linear(input, self.weight, self.bias)
RuntimeError: CUDA error: CUBLAS_STATUS_NOT_INITIALIZED when calling `cublasCreate(handle)`
```

### CODE REVIEW

File: components/_impl/workers/subprocess_worker.py [hunk 1]
```diff
@@ -372,4 +378,5 @@ def __del__(self) -> None:
         self._stderr_f.close()
 
         # Finally, make sure we don't leak any files.
```
Comment (aaronenyeshi) [inline components/_impl/workers/subprocess_worker.py:380]: nit: could update this comment to reflect the change.

File: components/_impl/workers/subprocess_worker.py [hunk 2]
```diff
@@ -49,12 +49,18 @@ def args(self) -> typing.List[str]:
     """
 
     _working_dir: str
+    _user_save_output: bool = False
     _alive: bool = False
     _bootstrap_timeout: int = 10  # seconds
 
-    def __init__(self, timeout: typing.Optional[float] = None, extra_env: typing.Optional[typing.Dict[str, str]]=None) -> None:
+    def __init__(self, timeout: typing.Optional[float] = None,
+                 extra_env: typing.Optional[typing.Dict[str, str]]=None,
```
Comment (aaronenyeshi) [inline components/_impl/workers/subprocess_worker.py:57]: to improve readability, we could use `from typing import Dict, Optional`, and use those directly. Also, seems inconsistently spacing before and after `=` in line 56-58.
```
self, timeout: Optional[float]=None,
extra_env: Optional[Dict[str, str]]=None,
save_output_dir: pathlib.Path=None) -> None:
```

File: userbenchmark/test_bench/regression_detector.py
Comment (aaronenyeshi) [inline userbenchmark/test_bench/regression_detector.py:1]: Why is this empty file needed?

**Review by aaronenyeshi — APPROVED** [review]
LGTM, very small nits and questions

File: userbenchmark/test_bench/regression_detector.py
Comment (xuzhao9) [reply userbenchmark/test_bench/regression_detector.py:1]: They will be used in the subsequent PR: https://github.com/pytorch/benchmark/pull/2054/files

File: components/_impl/workers/subprocess_worker.py (hunk 1 above)
Comment (xuzhao9) [reply components/_impl/workers/subprocess_worker.py]: Thanks! Addressed

File: components/_impl/workers/subprocess_worker.py (hunk 2 above)
Comment (xuzhao9) [reply components/_impl/workers/subprocess_worker.py]: Addressed the formatting issue

**Pull request closed without merging**